RetroAchievements/RAWeb
# feat(settings): migrate from legacy PHP to React

I apologize for how large the diff is. While I reasonably tried to break out the changes into multiple PRs (back-end, dependencies, etc), this page just has many many editable forms which also required several added base components. It's a lot of concepts and patterns all wrapped up into one package, virtually none of which can be executed in isolation (a server-side React component cannot be rendered into a PHP file).

---

This PR rewrites the user settings page in React and is the culmination of work done in https://github.com/RetroAchievements/RAWeb/pull/2579, https://github.com/RetroAchievements/RAWeb/pull/2596, and https://github.com/RetroAchievements/RAWeb/pull/2621. 

## How to test this PR
If you just want to take a quick tour, check out https://preview.retroachievements.org/settings.

To run locally:
```
sail npm run build
sail artisan inertia:start-ssr
```
Now you can navigate to http://localhost:64000/settings and start poking around.
The original route http://localhost:64000/controlpanel.php is still there to compare functionality. That file can be deleted in a quick follow-up.

There are virtually no back-end changes in this branch. It is 99.9% front-end, with the exception of adding `UserSettingsController::show()`.

## UX Walkthrough

### At a glance

The new page works great on both desktop and mobile. This is important, as 1 in 4 viewers on the page are [confirmed to be on mobile](https://plausible.retroachievements.org/retroachievements.org?auth=aMQAQjr37_a8AlTwzMXeN&filters=((is,page,(/controlpanel.php)))).

![Screenshot 2024-08-18 at 10 12 41 AM](https://github.com/user-attachments/assets/8917c3f9-5d2e-47cd-88ad-15c310b7c0e6)

![Screenshot 2024-08-18 at 10 12 07 AM](https://github.com/user-attachments/assets/7ed7272a-8d44-4a19-872c-622e6462295f)

When the user submits a form, it hits one of the controller endpoints built out in https://github.com/RetroAchievements/RAWeb/pull/2596. The user sees a helpful toast notification, and the page does not refresh:


https://github.com/user-attachments/assets/ae8771a9-b0f8-47e7-965a-43d4988b45ef

### Notable Changes

* For accessibility reasons, most forms now _require_ the user to press some kind of submit button to update data.
* In the Profile section, there is a "Visible Role" field. Currently it's set as read-only text to the user's stringified legacy `Permissions` value. Eventually, this should map to the new site roles we've been building out, with the possibility of allowing the user to select which role they'd like to have prominently displayed on their profile, user card, and forum posts:

![Screenshot 2024-08-18 at 10 20 45 AM](https://github.com/user-attachments/assets/68dc7bee-aa19-48a8-a4b9-ba09a155f308)

* In the Notifications section, the weekly RA newsletter option has been removed.
* Preferences are now driven by switches than checkboxes (although, they are still checkboxes in the DOM).
* In the Keys section, we no longer display the user's entire web API key. This is done so that if the user is screen recording, streaming, etc, they don't accidentally expose their web API key:

![Screenshot 2024-08-18 at 10 23 59 AM](https://github.com/user-attachments/assets/8b76af01-fa94-4ace-a6ca-852debc09873)

* In the Reset Game Progress section, the `<BaseSelect />` component is used rather than a native select. This prevents the select options from filling the entire height of the screen on some operating systems.

![Screenshot 2024-08-18 at 10 25 20 AM](https://github.com/user-attachments/assets/a2658269-ee5b-42c1-9f66-76bca288d0b3)

## How the Code Works

The root component of the page is _resources/js/pages/settings.tsx_.

Each section of the page is broken up logically into a separate & isolated component.

For the purpose of this walkthrough, we'll pick something simple: _ChangeEmailAddressSectionCard.tsx_. This is the component responsible for the "Change Email" section on the page:

![Screenshot 2024-08-18 at 10 28 29 AM](https://github.com/user-attachments/assets/c4b1f1be-6986-46e6-871a-cd83cec992cb)

Note that this component has nearly 100% code coverage via _ChangeEmailAddressSectionCard.test.tsx_ to verify its functionality.

The *.tsx file renders the template, of which we'll see several form fields, labels, etc.

The actual UI logic for submitting the data, handling validation, etc can be found within _useChangeEmailAddressForm.ts_:

```ts
const { form, mutation, onSubmit } = useChangeEmailAddressForm({ setCurrentEmailAddress });
```

In this React hook file, we first create the form schema using [zod](https://zod.dev/):
```ts
const changeEmailAddressFormSchema = z
  .object({
    newEmail: z.string().email(),
    confirmEmail: z.string().email(),
  })
  .refine((data) => data.newEmail === data.confirmEmail, {
    message: 'Email addresses must match.',
    path: ['confirmEmail'],
  });
```

The form schema is then input into our type-safe `useForm()` hook, powered by [react-hook-form](https://react-hook-form.com/):
```ts
const form = useForm<FormValues>({
  resolver: zodResolver(changeEmailAddressFormSchema),
  defaultValues: {
    newEmail: '',
    confirmEmail: '',
  },
});
```

All client-side AJAX handling is managed by [react-query](https://tanstack.com/query/latest). In this case, we're going to be sending some data to the server, so we use react-query's `useMutation()` hook:
```ts
const mutation = useMutation({
  mutationFn: (formValues: FormValues) => {
    return axios.put(route('settings.email.update'), formValues); // note our use of type-safe routes here
  },
  onSuccess: () => {
    props.setCurrentEmailAddress(form.getValues().newEmail);
  },
});
```

When the form indicates the user is submitting, we call `mutation.mutateAsync(formValues)` to send the data to the Laravel back-end:
```ts
const onSubmit = (formValues: FormValues) => {
  const confirmationMessage = auth?.user.roles.length
    ? 'Changing your email address will revoke your privileges and you will need to have them restored by staff. Are you sure you want to do this?'
    : 'Are you sure you want to change your email address?';

  if (!confirm(confirmationMessage)) {
    return;
  }

  toastMessage.promise(mutation.mutateAsync(formValues), {
    loading: 'Changing email address...',
    success: 'Changed email address!',
    error: 'Something went wrong.',
  });
};
```

This pattern is used in all of the forms on the page.

## Follow-up Stuff

In subsequent PRs, I will:
* Remove _controlpanel.blade.php_.
* Migrate all hardcoded strings to Laravel i18n resource JSON files, so we can support internationalization on this page and have an established pattern for using those translatable strings in React components.
* Increase the test coverage for the reset progress stuff. There are a few `it.todo()`s in _ResetGameProgressSectionCard.test.tsx_ that will require quite a bit of effort to fill in.
